В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [60]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [61]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [62]:
mean = np.mean(heights)
null_mean = 170
sd = np.std(heights)
alpha = 0.05

In [63]:
print("Середнє вибірки: {:.2f}".format(mean))
print("Стандартне відхилення вибірки: {:.2f}".format(sd))

Середнє вибірки: 170.70
Стандартне відхилення вибірки: 4.65


In [64]:
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='larger')
ztest_Score, p_value

(np.float64(0.8482235371587619), np.float64(0.19815674571091746))

In [65]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**ВИСНОВОК**

H0: середній зріст випадкової вибірки = 170 см

H1: середній зріст випадкової вибірки більший 170 см

Середній зріст у вибірці становить приблизно 170,70 см, тобто він чисельно вищий за 170 см. Однак за результатом одностороннього Z-тесту `p-value` = 0.198, що більше за рівень значущості 0.05. Отже, результат Z-тесту свідчить про те, що відхилити Н0 (прийняти Н1) не можна.  Недостатньо статистичних доказів, щоб стверджувати, що середній зріст студентів генеральної сукупності перевищує 170 см.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [66]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [67]:
print('Середнє групи A:', np.mean(group_a_scores))
print('Середнє групи B:', np.mean(group_b_scores))

Середнє групи A: 77.08125
Середнє групи B: 74.73875000000001


In [68]:
ztest_Score, p_value = ztest(group_a_scores, group_b_scores, value=0, alternative='two-sided')
ztest_Score, p_value

(np.float64(2.357354864751502), np.float64(0.01840564855252801))

In [69]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


**ВИСНОВОК**

H0: середня оцінка групи з новим курсом і групи зі стандартним ***однакова*** (іншими словами, різниця між середніми = 0)

H1: середня оцінка групи з новим курсом і групи зі стандартним ***різниться***

Середній оцінка у вибірці А (з новим курсом) становить приблизно 77,08, в В групі (зі стандартним курсом) - 74,7. Тобто на основі значень середніх група з новим курсом отримує вищі оцінки в середньому.

За результатом двостороннього Z-тесту p-value = 0.018, що менше за рівень значущості 0.05. Отже, результат Z-тесту свідчить про те, що можна відхилити Н0. Тобто прийняти Н1 і зробити висновок, що новий курс ефективніший.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [70]:
from scipy.stats import ttest_ind

In [71]:
# вибірка тільки перших 5 результатів групи A
group_a_5 = group_a_scores[:5]

In [72]:
print('Кількість записів у групі A:', len(group_a_5))
print('Кількість записів у групі B:', len(group_b_scores))

Кількість записів у групі A: 5
Кількість записів у групі B: 40


In [73]:
print('Середнє групи A:', np.mean(group_a_5))
print('Середнє групи B:', np.mean(group_b_scores))

Середнє групи A: 76.444
Середнє групи B: 74.73875000000001


In [74]:
t_score, p_value = ttest_ind(
    group_a_5,
    group_b_scores,
    equal_var=False, # t-тест Велча, який не вимагає припущення про однакові дисперсії груп (за замовченням True і припускаємо зворотнє)
    alternative='two-sided')

In [75]:
print('T-statistic:', t_score)
print('P-value:', p_value)

T-statistic: 1.1226527301193456
P-value: 0.3010736990344312


In [76]:
if p_value < alpha:
    print('Відхиляємо H0.')
else:
    print('Не відхиляємо H0.')

Не відхиляємо H0.


**ВИСНОВОК**

H0: середня оцінка групи з новим курсом і групи зі стандартним ***однакова*** (іншими словами, різниця між середніми = 0)

H1: середня оцінка групи з новим курсом і групи зі стандартним ***різниться***

Для групи A було використано лише перші 5 спостережень, тоді як для групи B - усі 40. Середня оцінка групи A становить приблизно 76,4, а групи B - 74,7.

За результатом двостороннього t-тесту Велча p-value = 0.301. Оскільки p-value перевищує рівень значущості 0.05, Н0 не відхиляємо. Недостатньо статистичних доказів, щоб стверджувати, що середні оцінки двох груп відрізняються.

Окремо варто зазначити, що порівняно з попереднім тестом статистична значущість зникла через зменшення вибірки групи A до 5 спостережень. Це не доводить, що новий курс неефективний, а лише означає, що такої малої вибірки недостатньо для переконливого підтвердження різниці.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [77]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

***T-test***

In [78]:
from scipy.stats import ttest_rel # t-test для двох повʼязаних вибірок

In [79]:
before = np.array(before_training)
after = np.array(after_training)

In [80]:
print('Кількість значень до:', len(before))
print('Кількість значень після:', len(after))

Кількість значень до: 30
Кількість значень після: 30


In [81]:
differences = after - before # різниця оцінок для кожного працівника

In [82]:
print(f'Середнє до тренінгу: {before.mean():.2f}')
print(f'Середнє після тренінгу: {after.mean():.2f}')
print(f'Середнє покращення/погіршення: {differences.mean():.2f}')

Середнє до тренінгу: 50.89
Середнє після тренінгу: 54.04
Середнє покращення/погіршення: 3.15


In [83]:
t_statistic, p_value = ttest_rel(
    after,
    before,
    alternative='greater' # greater, бо цікавить саме напрям зміни (тобто after > before, тоді результат різниці буде додатнім), а не просто чи відрізняються середні
    )

print(f'T-статистика: {t_statistic:.3f}')
print(f'P-value: {p_value:.10f}')

T-статистика: 9.139
P-value: 0.0000000002


In [84]:
if p_value < alpha:
    print('Відхиляємо H0.')
else:
    print('Не відхиляємо H0.')

Відхиляємо H0.


**ВИСНОВОК**

H0: середня оцінка працівників до тренінгу та після тренінгу (працівник в обох вибірках) ***однакова*** (іншими словами, різниця між середніми = 0)

H1: середня оцінка працівників до тренінгу та після тренінгу ***різниться***

Середній показник до тренінгу становить приблизно 50.9, а після тренінгу - 54. Середнє покращення становить 3.15 одиниці.

За результатом одностороннього парного t-тесту отримано p-value = 0.000... Оскільки p-value значно менше за рівень значущості 0.05, Н0 відхиляємо на користь альтернативної. Отже, дані свідчать про статистично значуще підвищення продуктивності працівників після тренінгу.

***Z-test***

In [89]:
# для того, щоб обійти повʼязані вибірки, порівнюємо різницю оцінок зі значенням 0
# якщо різниці немає тобто after - before = 0, то оцінки однакові, якщо after - before > 0, то after > before

z_statistic, p_value = ztest(differences, value=0, alternative='larger')

print(f'Z-статистика: {z_statistic:.3f}')
print(f'P-value: {p_value:.10e}')

Z-статистика: 9.139
P-value: 3.1545359919e-20


In [86]:
if p_value < alpha:
    print('Відхиляємо H0.')
else:
    print('Не відхиляємо H0.')

Відхиляємо H0.


**ВИСНОВОК**

H0: середня різниця між оцінками до і після тренінгу = 0 (тобто оцінки однакові)

H1: середня різниця між оцінками до і після тренінгу > 0 (тобто середня оцінка після тренінгу вища оцінки до)

За результатом одностороннього Z-тесту `p-value` значно нижчий за рівень значущості 0.05. Отже, можена відхилити Н0 (прийняти Н1). Таким чином, середня оцінка після тренінгу вища за середню оцінку до тренінгу.